File: pilha/pilha.py
class Pilha:
  def __init__(self, maxelem):
    self.pilha = [0] * maxelem
    self.tamanho = 0

  def empilhar(self, x):
    i = 0
    while i < self.tamanho and self.pilha[i] != x:
      i += 1
    if i == self.tamanho:
      if self.tamanho < len(self.pilha):
        self.pilha[self.tamanho] = x
        self.tamanho += 1
        return True
      else:
        return False
    else:
      return False

  def desempilhar(self):
    if self.tamanho == 0:
      return False
    
    elemento = self.pilha[self.tamanho-1]
    self.tamanho -= 1
    return elemento

  def imprimir(self):
    print(self.pilha[:self.tamanho])
  
pilha = Pilha(5)
pilha.empilhar(1)
pilha.empilhar(2)
pilha.imprimir()

pilha.desempilhar()
pilha.imprimir()
    

File: estrutura-de-dados-em-java/src/pilha/Pilha.java
package pilha;

public class Pilha {
  private String[] elementos;
  private int tamanho;

  public Pilha(int capacidade){
    this.elementos = new String[capacidade];
    this.tamanho = 0;
  }

  public void empilhar(String elemento){
    this.aumentarCapacidade();

    int i = 0;
    while (i < this.tamanho && this.elementos[i] != elemento) {
      i++;
    }
    if (i == this.tamanho) {
      if (this.tamanho < this.elementos.length) {
        this.elementos[this.tamanho] = elemento;
        this.tamanho++;
      }
    }
  }

  public boolean estaVazia(){
    return this.tamanho == 0;
  }

  public String topo(){
    return this.elementos[this.tamanho - 1];
  }

  public String desempilhar(){
   if(this.estaVazia()){
     return null;
   }

   return this.elementos[--tamanho];
  }

  private void aumentarCapacidade(){
    if (this.tamanho == this.elementos.length) {
      String[] elementosNovos = new String[this.elementos.length * 2];
      int i = 0;
      while (i < this.elementos.length) {
        elementosNovos[i] = this.elementos[i];
        i++;
      }
      this.elementos = elementosNovos;

      System.out.println("Aumentou a capacidade para " + this.elementos.length);
    }
  }

  @Override
  public String toString() {
    StringBuilder s = new StringBuilder();
    s.append("[");
    int i = 0;

    while (i < this.tamanho - 1) {
      s.append(this.elementos[i]);
      s.append(", ");
      i++;
    }

    if (this.tamanho > 0) {
      s.append(this.elementos[this.tamanho - 1]);
    }

    s.append("]");
    return s.toString();
  }
}
